### ISSUE TITLE
Comments are not considered part of the block

### ISSUE THREAD

**Issue by dionysius** [bombsimon/wsl#62]
Hi, I found a part of my code which wsl should have reported linting errors, but hasn't because the comments were the "block separator". Consider this arbitary file:

```golang
package root

import "fmt"

func init() {
	var foo string
	// some comment
	var bar string
	// if something
	err := fmt.Errorf("foo")
	if err != nil {
		fmt.Println("do-something")
		// bla
		return
	}
	// something else
	var baz string
	// and
	return
}
```

I would've expected that different wsl rules would be reported, but aren't.

```
$ go run github.com/bombsimon/wsl/cmd/wsl -w test.go
<no output>
$ go list -json github.com/bombsimon/wsl/cmd/wsl | jq -r '.Module.Version'
v1.2.8
```

I would've expected that the comments are part of that block, and a newline before the comment fulfills the separation.

for comparison this would satisfy wsl if there were no comments

```golang
package root

import "fmt"

func init() {
	var foo string

	// some comment
	var bar string

	// if something
	err := fmt.Errorf("foo")
	if err != nil {
		fmt.Println("do-something")

		// bla
		return
	}

	// something else
	var baz string

	// and
	return
}
```

**Comment by bombsimon** [comment]
Thanks for the report!

This is "kind of" intentional. I've known about this since the start and I haven't made up my mind if this is how I want it or not. Since the whole idea with this linter is to improve readability which I think is done by separating with newliens **or** comments I think I'm ok with this.

The same goes for leading comments. The linter doesn't allow you to start a block with a newline but it may start with comments. That's not because I see comment's as statements or nodes, it's because I don't think it has a bad impact on the readability.

However, with the work on #4 I need to take comments for a statement into consideration so maybe I should bundle each node with it's leading and trailing comments when they're found and have this configurable.

I'll definitely leave this ticket open, I hope people feel that this is the right way gives this a thumbs up. Depending on the work with grouping statements and comments together this might be an easy implementation in a future release.

**Comment by mholiday-nyt** [comment]
I find it odd that this fails:

```go
func xyz() {
    // here's a starting comment with whitespace after

    if x == nil {
    }

    . . .
```

because ignoring the comment as an element then causes the whitespace after it to be "leading whitespace" within the block

**Comment by bombsimon** [comment]
Yes that doesn't sound right, I wonder if it's a result of mergin #73 or if it was this way before. I don't think that should be valid but working with comments, especially those not obvious bound to a node, is the trickiest part and one of the reasons I still not finished the fixer for wsl. Thisi one should be fairly simple though 🤞🏼

**Comment by mholiday-nyt** [comment]
I've found what I think is the fix for the above, I'll open another issue and PR for it

        // If we have multiple groups, add support for newline between each group.
        if p.config.AllowSeparatedLeadingComment {
-               if seenCommentGroups > 1 {
-                       allowedLinesBeforeFirstStatement += seenCommentGroups - 1
+               if seenCommentGroups > 0 {
+                       allowedLinesBeforeFirstStatement += seenCommentGroups
                }
        }

and to add the missing doc for that flag :0)

**Comment by bombsimon** [comment]
Nice find! Yeah that must be it, at least if one should trust the comment. :)

**Comment by mholiday-nyt** [comment]
Will continue the discussion in #104; that logic is not quite good enough :-(

**Comment by breml** [comment]
I would like to add one though to the discussion about comments being seen similar (or even the same) as newlines as stated by @bombsimon in https://github.com/bombsimon/wsl/issues/62#issuecomment-559075732

> Since the whole idea with this linter is to improve readability which I think is done by separating with newliens or comments I think I'm ok with this.

I remember a conversation on a podcast (unfortunately I am not be able to dig out the original source of this), where a blind person, who works as a software developer, explained how it is important for them, that code has a good structure in the sense of being split into "paragraphs" of code, which does belong together.
For someone, who can see normally, a comment might provide a good way to structure parts visually, in particular if the comments have a different color as it is the case in most modern editors or IDE. But for a blind person, this is not similarly the case and it might even hinder the person to navigate quickly through the code, if the code is not split into "paragraphs" separated by empty lines.

**Comment by bombsimon** [comment]
> I got the impression, that in particular `/* */` comments are difficult

There are several issues with comments and that's one of them, but the fact that it's not a part of the regular AST but has its own makes it complex and annoying. Iterating over all comments in the file and look for positions like you do in `newline-after-block` doesn't feel ideal to me since it's not very efficient.

`wsl` have the cursor concept where we can iterate over all statements. That would have to be rewritten independent of comment type type not only look for next statement but also look for next comment. Which doesn't have to be an issue, just something I've avoided.

> I would even go so far, that I would be fine, if a linter would only consider line comments for this regard.

I'm torn on this approach. In general having caveats like that feels wrong and I try to support all cases and not just part or specific ones. I don't think `wsl` has a huge audience but it feels hard to argue why one typ of comment would work but not the other.

I also don't think i would be any easier, they're represented the same way in the comment AST.

> If you have some interesting samples with hard to process comments, it would be interesting to see

It's basically for any comment to figure out where the start and end, which group they belong to etc.

Given that `wsl` does the analysis from the statement _after_ where we want the newline, we would need to first check if the line above is another statement as we do today. But if not, we would need to see if it's a comment. Doing that require us to fetch comments via the comment map for either the current or previous statement (due to linked issue with free floating comments). If it is, we must find the start of the comment and see if _that_ is directly after a statement. If it's a comment group, we would need to find the first comment in the group and see where _that_ starts. 

So although not complex comment, something like this would require such flow for `wsl` to support all current rules but with an option to not treat comments as empty lines.

```go
go func() {}
// Comment group comment[0]

// Comment group comment[1]
fmt.Println("")
```

It's also a bit hard to reason around, if that's allowed (due to the whitespace) why would not these two be?

```go
go func() {}
// Comment group comment[0]
//
// Comment group also comment[0]
fmt.Println("")
```

```go
go func() {}
/* Comment group comment[0]

Comment group also comment[0] */
fmt.Println("")
```

And this would only be one comment group and thus not allowed

```go
go func() {}
// Comment group comment[0]
// Comment group also comment[0]
fmt.Println("")
```

But this one would be two, so it would feel odd to me to allow it

```go
go func() {}
/* Comment group comment[0] */
/* Comment group comment[1] */
fmt.Println("")
```

I see `newline-after-block` uses a completely different approach to find the `Pos` for diagnostic via the `findEndOfLine` with `LineStart` which I wasn't aware of. I suspect something like that could simplify `wsl` code quite a lot. I'll have to look into that.

I haven't considered this for a long time but I can do some investigation to see how much effort and how it would effect the code base to support this.

**Comment by breml** [comment]
> ```golang
> go func() {}
> // Comment group comment[0]
> 
> // Comment group comment[1]
> fmt.Println("")
> ```
> It's also a bit hard to reason around, if that's allowed (due to the whitespace) why would not these two be?

Me personally, I tend to put the comments before the thing the comments relate to. So the question is, what is `Comment group comment[0] about. As the automatic fix, I would expect the linter to propose the following:

```golang
go func() {}

// Comment group comment[0]

// Comment group comment[1]
fmt.Println("")
```

Obviously, the author of the code is then free to join the comments, if this is what matches their intention. But then the code would need to look like this (since it follows a block):

```golang
go func() {}

// Comment group comment[0]
//
// Comment group comment[1]
fmt.Println("")
```

> ```golang
> go func() {}
> /* Comment group comment[0] */
> /* Comment group comment[1] */
> fmt.Println("")
> ```

This is an interesting example and I wonder, if `wsl` maybe should even be opinionated about block comments taking a complete line and therefore proposing:

```golang
go func() {}

// Comment group comment[0]
// Comment group comment[1]
fmt.Println("")
```

**Comment by breml** [comment]
Just fund, that the examples are not even valid Go code, since `go func() {}` needs to be `go func() {}()` and then it is no longer a block (but a function call), which does not require a newline after it. So the example would need to be: `_ = func() {}` (or similar).
